crates/fluvio/src/consumer/retry.rs

```diff
@@ -0,0 +1,279 @@
+        }
+    }
+}
+
+impl ConsumerWithRetry {
+    /// Creates a new `ConsumerWithRetry` instance.
+    pub async fn new(fluvio_config: FluvioConfig, config: ConsumerConfigExt) -> Result<Self> {
+        let fluvio = Fluvio::connect_with_config(&fluvio_config).await?;
+        let stream = fluvio.consumer_with_config_inner(config.clone()).await?;
+        let boxed_stream: BoxConsumerStream = Box::pin(stream);
+
+        Ok(Self {
+            inner: ConsumerWithRetryInner {
+                fluvio_client: Arc::new(fluvio),
+                next_offset_to_read: None,
+                consumer_config: config,
+            },
+            current_stream: Some(boxed_stream),
+            pending: None,
+        })
+    }
+
+    /// Consumes records from the stream with retry logic.
+    ///
+    /// This function continuously attempts to retrieve a record from the current stream.
+    /// On failure (or end-of-stream), it waits using an exponential backoff strategy
+    /// and then attempts to reconnect.
+    async fn consumer_with_retry(
+        inner: ConsumerWithRetryInner,
+        fluvio_client: Arc<Fluvio>,
+        mut current_stream: BoxConsumerStream,
+    ) -> Option<Result<(ConsumerRecord, BoxConsumerStream, Option<i64>), ErrorCode>> {
+        let mut attempts: u32 = 0;
+        let mut backoff = create_backoff();
+
+        loop {
+            // Try to retrieve the next record.
+            if let Some(record_result) = current_stream.as_mut().next().await {
+                match record_result {
+                    Ok(record) => {
+                        let new_offset = Some(record.offset + 1);
+                        if attempts > 0 {
+                            debug!(
+                                target: SPAN_RETRY,
+                                "Record produced successfully after reconnect"
+                            );
+                        }
+                        return Some(Ok((record, current_stream, new_offset)));
+                    }
+                    Err(e) => {
+                        warn!(target: SPAN_RETRY, "Error consuming record: {}", e);
+                        if let RetryMode::Disabled = inner.consumer_config.retry_mode {
+                            return Some(Err(e));
+                        }
+                    }
+                }
+            }
+
+            // If continuous consumption is disabled, end the stream.
+            if inner.consumer_config.disable_continuous {
+                return None;
+            }
+
+            // Wait before retrying.
+            backoff_and_wait(&mut backoff).await;
+            attempts += 1;
+
+            // Determine the offset for reconnection.
+            let offset = if let Some(next) = inner.next_offset_to_read {
+                match Offset::absolute(next) {
+                    Ok(off) => off,
+                    Err(e) => {
+                        warn!(target: SPAN_RETRY, "Error creating offset: {}", e);
```

Comment:
return error?
or alternatively, have option to ignore error

Reply:
We're returning error at next line, no?

Reply:
k